### PR TITLE
Rewrite autoloader so we have all rules defined in separate class

### PR DESCRIPTION
**Pull request by michalbundyra**
The class with defined rules now is reusable and we can use it also on rewriting the libraries

~We need to think what we can do with functions, and somehow I think it will be better to create these alias in repositories where we have defined functions.~

~We would need define list of all possible function defined in libraries and then just call new function using `call_user_func_array` and `func_get_args`.~

**UPDATE:** As discussed, we will try to handle functions separately on rewriting packages, as these are only in the following pacakges:
- zend-diactoros
- zend-stragitility
- zend-container-config-test

### CODE REVIEW

File: src/Autoloader.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+            $i = 0;
+            $check = '';
+
+            while (isset($classes[$check . $segments[$i] . '\\'])) {
+                $check .= $segments[$i] . '\\';
+                ++$i;
+            }
+
+            if ($check !== '') {
```
Comment (weierophinney) [inline src/Autoloader.php]: I'd invert this condition:

```php
if ($check === '') {
    return;
}
```

File: src/Autoloader.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        spl_autoload_register(static function ($class) use ($classes) {
+            $segments = explode('\\', $class);
+
+            $i = 0;
+            $check = '';
+
+            while (isset($classes[$check . $segments[$i] . '\\'])) {
+                $check .= $segments[$i] . '\\';
+                ++$i;
+            }
```
Comment (weierophinney) [inline src/Autoloader.php:31]: This seems really convoluted, honestly. Why not just:

```php
foreach ($class as $legacyPrefix => $aliasPrefix) {
    if (strpos($class, $legacyPrefix) === 0) {
        $alias = str_replace($legacyPrefix, $aliasPrefix, $class);
        class_alias($alias, $class);
        return;
    }
}
```

**Review by weierophinney — CHANGES_REQUESTED** [review]

File: src/RewriteRules.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?php
+/**
+ * @see       https://github.com/laminas/laminas-zendframework-bridge for the canonical source repository
+ * @copyright Copyright (c) 2019 Laminas Foundation (https://getlaminas.org)
```
Comment (michalbundyra) [inline src/RewriteRules.php]: we should update it now to refer to COPYRIGHT.md file now

File: src/Autoloader.php (hunk 2 above)
Comment (michalbundyra) [reply src/Autoloader.php:31]: It was exactly like that before. Now we are using segments, so we can return quicker - in 1 or two steps. With foreach we need to check it in the worst case scenario 12 times...

Comment (weierophinney) [reply src/Autoloader.php:31]: Yep - I was misreading it previously.

Maybe put in a comment about how it works, so others who come to it later know why it's written the way it is.

File: src/RewriteRules.php (hunk 3 above)
Comment (weierophinney) [reply src/RewriteRules.php]: Yes, please. :)

File: src/Autoloader.php (hunk 2 above)
Comment (michalbundyra) [reply src/Autoloader.php:31]: Comment is added now.

**Pull request merged**